Client/mods/deathmatch/logic/luadefs/CLuaEngineDefs.cpp

```diff
@@ -2560,3 +2591,188 @@ void CLuaEngineDefs::EnginePreloadWorldArea(CVector position, std::optional<Prel
+    if (!modelInfo)
+        return false;
+
+    auto* effect = modelInfo->Get2DFXFromIndex(index);
+    if (!effect)
+        return false;
+
+    const char* error = CClient2DFXManager::IsValidEffectData(effect->type, effectData);
+    if (error)
+        throw std::invalid_argument(error);
+
+    modelInfo->StoreDefault2DFXEffect(effect);
+    return m_p2DFXManager->Set2DFXProperties(effect, effectData);
+}
+
+bool CLuaEngineDefs::EngineSet2DFXProperties(CClient2DFX* effect, std::unordered_map<std::string, std::variant<bool, float, std::string>> effectData)
+{
+    const char* error = CClient2DFXManager::IsValidEffectData(effect->Get2DFXType(), effectData);
+    if (error)
+        throw std::invalid_argument(error);
+
+    return m_p2DFXManager->Set2DFXProperties(effect->Get2DFX(), effectData);
+}
+
+bool CLuaEngineDefs::EngineSetModel2DFXPosition(std::uint32_t modelID, std::uint32_t index, CVector position)
+{
+    if (!CClient2DFXManager::IsValidModel(modelID))
+        throw std::invalid_argument("Invalid model ID");
+
+    auto count = EngineGetModel2DFXCount(modelID);
+    if (std::holds_alternative<std::uint32_t>(count) && index >= std::get<std::uint32_t>(count))
+        throw std::invalid_argument("Invalid effect index");
+
+    CModelInfo* modelInfo = g_pGame->GetModelInfo(static_cast<DWORD>(modelID));
+    if (!modelInfo)
+        return false;
+
+    auto* effect = modelInfo->Get2DFXFromIndex(index);
+    if (!effect)
+        return false;
+
+    modelInfo->StoreDefault2DFXEffect(effect);
+    m_p2DFXManager->Set2DFXPosition(effect, position);
+    return true;
+}
+
+bool CLuaEngineDefs::EngineSet2DFXPosition(CClient2DFX* effect, CVector position)
+{
+    m_p2DFXManager->Set2DFXPosition(effect->Get2DFX(), position);
+    return true;
+}
+
+std::variant<bool, std::tuple<float, float, float>> CLuaEngineDefs::EngineGetModel2DFXPosition(std::uint32_t modelID, std::uint32_t index)
+{
+    if (!CClient2DFXManager::IsValidModel(modelID))
+        throw std::invalid_argument("Invalid model ID");
+
+    auto count = EngineGetModel2DFXCount(modelID);
+    if (std::holds_alternative<std::uint32_t>(count) && index >= std::get<std::uint32_t>(count))
+        throw std::invalid_argument("Invalid effect index");
+
+    CModelInfo* modelInfo = g_pGame->GetModelInfo(static_cast<DWORD>(modelID));
+    if (!modelInfo)
+        return false;
+
+    auto* effect = modelInfo->Get2DFXFromIndex(index);
+    if (!effect)
+        return false;
+
+    CVector* position = m_p2DFXManager->Get2DFXPosition(effect);
+    return std::make_tuple(position->fX, position->fY, position->fZ);
+}
+
+std::variant<bool, std::tuple<float, float, float>> CLuaEngineDefs::EngineGet2DFXPosition(CClient2DFX* effect)
+{
+    CVector* position = m_p2DFXManager->Get2DFXPosition(effect->Get2DFX());
+    if (!position)
+        return false;
+
+    return std::make_tuple(position->fX, position->fY, position->fZ);
+}
```

Comment:
use CLuaMultiReturn?

Reply:
This is exactly the same, so it makes no difference https://github.com/multitheftauto/mtasa-blue/blob/81c939a4e047378d166da28dfce7dfa85bdc368d/Shared/mods/deathmatch/logic/lua/CLuaMultiReturn.h